### PR TITLE
Add arrayPartialSort and arrayPartialReverseSort

### LINKED ISSUE #11125: random permutation/selection for arrays
Implement functions for generating random permutations and random selections of an array.

I need permutations for generating randomization distribution in performance tests, and now have to emulate it with `order by rand()` which becomes painfully slow for big arrays. In contrast, shuffle algorithms work in linear time.

Selections would also be useful in performance tests for bootstrapping confidence intervals for query run time.

### PR DESCRIPTION
**Pull request by jh0x**
Closes #11125

- Modified `FunctionArrayMapped` to allow for extra positional arguments in mapped functions (i.e. `arrayPartialSort(limit, arr)` or `arrayPartialSort(x -> predicate, limit, arr)`.
- Added `arrayPartialSort` and `arrayPartialReverseSort`

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added arrayPartialSort and arrayPartialReverseSort functions

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by jh0x** [comment]
I will investigate the failures.

File: src/Functions/array/FunctionArrayMapped.h [hunk 1]
```diff
@@ -144,32 +161,54 @@ class FunctionArrayMapped : public IFunction
...
+                    "The {}{}{} argument for function {} must be array. Found {} instead",
+                    num_fixed_params + 1,
+                    getOrdinalSuffix(num_fixed_params + 1),
+                    (num_fixed_params == 0 ? " and only" : ""),
+                    getName(),
+                    arguments[num_fixed_params].type->getName());
+
+            if constexpr (num_fixed_params)
+                Impl::checkArguments(
+                    std::span<const ColumnWithTypeAndName, num_fixed_params>(std::begin(arguments), num_fixed_params), getName());
```
Comment (yakov-olkhovskiy) [inline src/Functions/array/FunctionArrayMapped.h:199]: here and in `execute` - why do we need to use `span` in this strange and confusing way? can we just pass type to `checkArguments` and limit to `execute`?

File: src/Functions/array/arraySort.cpp [hunk 2]
```diff
@@ -53,25 +88,106 @@ struct ArraySortImpl
...
-}
 
+    factory.registerFunction<FunctionArrayPartialSort>({
+        R"(
+Returns an array of the same size as the original array where elements in range `[1..limit]`
+are sorted in ascending order. Remaining elements `(limit..N]` shall contain elements in unspecified order.
+[example:simple_int]
+[example:simple_string]
+
+If one wishes to retain only the sorted elements it can be achieved with:
```
Comment (yakov-olkhovskiy) [inline src/Functions/array/arraySort.cpp:144]:
```suggestion
To retain only the sorted elements use `arrayResize`:
```

File: src/Functions/array/FunctionArrayMapped.h [hunk 3]
```diff
@@ -144,32 +161,54 @@ class FunctionArrayMapped : public IFunction
...
+                nested_types.size(),
+                arguments[0]->getName());
 
         arguments[0] = std::make_shared<DataTypeFunction>(nested_types);
     }
 
     DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
     {
-        size_t min_args = Impl::needExpression() ? 2 : 1;
+        size_t min_args = (num_fixed_params + Impl::needExpression()) ? 2 : 1;
```
[the remark this answers is not in the document]
Comment (jh0x) [reply src/Functions/array/FunctionArrayMapped.h]: Good catch, fixed!

File: src/Functions/array/FunctionArrayMapped.h (hunk 1 above)
Comment (yakov-olkhovskiy) [reply src/Functions/array/FunctionArrayMapped.h]: but fixed params always have the same type `ColumnWithTypeAndName` - why do we need to unwind arguments into tuple? we can just pass iterator on the first fixed argument, and then `checkArguments` and `execute` knows the number of such params... I don't see the point of such complications

**Review by yakov-olkhovskiy — APPROVED** [review]

**Pull request merged**